picard/ui/tablebaseddialog.py

```diff
@@ -0,0 +1,200 @@
+        self.sorting_enabled = True
+        self.create_table()
+        self.finished.connect(self.save_state)
+
+    @property
+    def columns(self):
+        return self.__columns
+
+    @columns.setter
+    def columns(self, list_of_tuples):
+        if not list_of_tuples:
+            list_of_tuples = []
+        self.__columns = OrderedDict(list_of_tuples)
+        self.__colkeys = list(self.columns.keys())
+
+    @property
+    def table_headers(self):
+        return list(self.columns.values())
+
+    def colpos(self, colname):
+        return self.__colkeys.index(colname)
+
+    @abstractmethod
+    def get_value_for_row_id(self, row, value):
+        pass
+
+    def set_table_item(self, row, colname, obj, key, sortkey=None):
+        value = obj.get(key, "")
+        self.set_table_item_val(row, colname, value, sortkey)
+
+    def set_table_item_val(self, row, colname, value, sortkey=None):
+        # QVariant remembers the original type of the data
+        # matching comparison operator will be used when sorting
+        # get() will return a string, force conversion if asked to
+        if sortkey is None:
+            sortkey = natsort.natkey(value)
+        item = SortableTableWidgetItem(sortkey)
+        item.setData(QtCore.Qt.DisplayRole, value)
+        pos = self.colpos(colname)
+        if pos == 0:
+            id = self.get_value_for_row_id(row, value)
+            item.setData(QtCore.Qt.UserRole, id)
+        self.table.setItem(row, pos, item)
+
+    @abstractmethod
+    def setupUi(self):
+        pass
+
+    def add_widget_to_center_layout(self, widget):
+        """Update center widget with new child. If child widget exists,
+        schedule it for deletion."""
+        widget_item = self.center_layout.takeAt(0)
+        if widget_item:
+            current_widget = widget_item.widget()
+            current_widget.hide()
+            self.center_layout.removeWidget(current_widget)
+            if current_widget != self.table:
+                current_widget.deleteLater()
+        self.center_layout.addWidget(widget)
+        widget.show()
+
+    def create_table_obj(self):
+        return ResultTable(self)
+
+    def create_table(self):
+        self.table = self.create_table_obj()
+        self.table.verticalHeader().setDefaultSectionSize(100)
+        self.table.setSortingEnabled(False)
+        self.table.cellDoubleClicked.connect(self.accept)
+        self.table.hide()
+
+        def enable_accept_button():
+            self.accept_button.setEnabled(True)
+        self.table.itemSelectionChanged.connect(enable_accept_button)
+
+    def prepare_table(self):
+        self.table.prepare(self.table_headers)
+        self.restore_table_header_state()
+
+    @abstractmethod
+    def show_table(self, sort_column=None, sort_order=QtCore.Qt.DescendingOrder):
+        pass
+
+    @abstractmethod
+    def accept(self):
+        super().accept()
```

Comment:
This is not really needed to implement here if it just calls `super().accept()`, since `QDialog` already provides this. So this implementation currently does not provide any value. But again I wonder whether we should include the current default implementation from searchdialog, e.g.:

```python
def accept(self):
    if self.table:
        selected_rows = []
        for idx in self.table.selectionModel().selectedRows():
            row = self.table.itemFromIndex(idx).data(QtCore.Qt.UserRole)
            selected_rows.append(row)
        self.accept_event(selected_rows)
    super().accept()

@abstractmethod
def accept_event():
    pass
```

This looks useful for me for a table based result dialog and your new match details dialog could then also just implement `accept_event`

Reply:
the super.accept() was required in the superclass, otherwise the dialog did not close.

I made the change, but in the overridden MatchDetailsDialog.accept() I had to call QDialog.accept(self) to get the dialog to close.

Reply:
That's strange, should not be needed. But I think the proper way would be to not reimplement `accept` in `MatchDetailsDialog` at all, but only `accept_event` and have it something like this:

```
    def accept_event(self, rows):
        if not rows:
            return
        selected_entry = rows[0]
        match_details_log.set_file_result_selected_entry(self.current_file.filename, selected_entry)
        self.load_selection(selected_entry)
```

That way the base class is generic and you only implement the special handling needed for `MatchDetailsDialog`. Yes, the generic implementation handles multiple rows, but it is easy to handle only a single selection in the subclass so this is fine.

Reply:
Done.  good idea!